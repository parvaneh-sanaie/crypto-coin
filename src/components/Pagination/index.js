import React, { useEffect, useState } from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import PropTypes from 'prop-types';

import calculatePaginationRecords from '../../utils/pagination';

function Pagination(props) {
    const { currentPage, size, totalCount, onPageChange, onPageSizeChange, className } = props;

    const MAX_DISPLAY_PAGES = 6;
    const [page, setPage] = useState(currentPage);
    const [pageSize, setPageSize] = useState(size);
    const [totalPages, setTotalPages] = useState(Math.floor(totalCount / size));
    const [displayPages, setDisplayPages] = useState(
        calculatePaginationRecords({ currentPage, maxDisplayPages: MAX_DISPLAY_PAGES, totalCount, size }),
    );

    useEffect(() => {
        setDisplayPages(
            calculatePaginationRecords({
                currentPage: page,
                maxDisplayPages: MAX_DISPLAY_PAGES,
                totalCount,
                size: pageSize,
            }),
        );
    }, [totalCount, page, pageSize, totalPages]);

    const handlePageChange = (pageNumber) => {
        onPageChange(pageNumber);
        setPage(pageNumber);
    };

    const handleClickPreviousPage = () => {
        onPageChange(page - 1);
        setPage(page - 1);
    };

    const handleClickNextPage = () => {
        if (page - 1 <= totalPages) {
            onPageChange(page + 1);
            setPage(page + 1);
        }
    };

    const handleChangePageSize = (event) => {
        const newPageSize = Number(event.target.value);
        onPageSizeChange(newPageSize);
        setPageSize(newPageSize);
        setTotalPages(Math.ceil(totalCount / newPageSize));
        setPage(1);
    };
    return (
        <div data-testid="pagination" className={`container flex-fill ${className}`}>
            <div className="row">
                <div className="col-xl-3 col-md-2 d-flex flex-fill justify-content-start">
                    <label data-testid="pagination-page-info">
                        {(page - 1) * pageSize + 1}-{totalCount < pageSize ? totalCount : page * pageSize} of{' '}
                        {totalCount}
                    </label>
                </div>
                <div className="col-xl-6 col-md-8 d-flex flex-fill justify-content-center">
                    <nav>
                        <ul data-testid="pagination-page-list" className="pagination">
                            <li
                                data-testid="pagination-previous-link"
                                className={`page-item ${page <= 1 ? 'disabled' : ''}`}
                            >
                                <button type="button" className="page-link" onClick={handleClickPreviousPage}>
                                    <span aria-hidden="true">&laquo;</span>
                                </button>
                            </li>
                            {displayPages.map((pageNumber) => {
                                const key = new Date().getMilliseconds() + Math.random();
                                return (
                                    <li className="page-item" key={`${key}`}>
                                        {pageNumber !== page && (
                                            <button
                                                type="button"
                                                className="page-link"
                                                onClick={() => handlePageChange(pageNumber)}
                                                data-page={pageNumber}
                                                key={`${key}${pageNumber}`}
                                            >
                                                <span aria-hidden="true">{pageNumber}</span>
                                            </button>
                                        )}
                                        {pageNumber === page && (
                                            <span key={`${key}${pageNumber}`} className="page-link active">
                                                {pageNumber}
                                            </span>
                                        )}
                                    </li>
                                );
                            })}
                            <li
                                data-testid="pagination-next-link"
                                className={`page-item ${page >= totalPages || totalCount < pageSize ? 'disabled' : ''}`}
                            >
                                <button type="button" className="page-link" onClick={handleClickNextPage}>
                                    <span aria-hidden="true">&raquo;</span>
                                </button>
                            </li>
                        </ul>
                    </nav>
                </div>
                <div className="col-xl-3 col-md-2 d-flex flex-fill justify-content-end">
                    <label>
                        Rows per page:
                        <select
                            data-testid="pagination-page-size-records"
                            value={pageSize}
                            onChange={handleChangePageSize}
                        >
                            <option value={5}>5</option>
                            <option value={10}>10</option>
                            <option value={20}>20</option>
                            <option value={50}>50</option>
                            <option value={100}>100</option>
                        </select>
                    </label>
                </div>
            </div>
        </div>
    );
}
Pagination.propTypes = {
    currentPage: PropTypes.number,
    size: PropTypes.number,
    totalCount: PropTypes.number,
    onPageChange: PropTypes.func,
    onPageSizeChange: PropTypes.func,
    className: PropTypes.string,
};
Pagination.defaultProps = {
    size: 10,
    currentPage: 1,
    totalCount: 1,
    onPageChange: (pageNumber) => pageNumber,
    onPageSizeChange: (size) => size,
    className: ``,
};
export default Pagination;
