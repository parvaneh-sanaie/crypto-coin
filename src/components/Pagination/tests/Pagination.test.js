import React from 'react';
import { cleanup, render, act, screen, within, fireEvent } from '@testing-library/react';

import Pagination from '../index';

describe('<Pagination />', () => {
    it('Should be match to the snapshot', () => {
        let renderResult = null;
        act(() => {
            renderResult = render(<Pagination />);
        });
        expect(renderResult).toMatchSnapshot();
    });

    it('should render the component with 1 page and disabled next and previous link', () => {
        expect.assertions(11);
        let renderResult = null;
        act(() => {
            renderResult = render(<Pagination />);
        });
        const pageList = screen.getByTestId('pagination-page-list');
        const nextLink = screen.getByTestId('pagination-next-link');
        const previousLink = screen.getByTestId('pagination-previous-link');
        expect(pageList).toBeInTheDocument();
        expect(nextLink).toBeInTheDocument();
        expect(previousLink).toBeInTheDocument();
        expect(screen.getByTestId('pagination-page-info').innerHTML).toBe('1-1 of 1');
        expect(renderResult.container.getElementsByClassName('active').length).toBeGreaterThanOrEqual(1);
        expect(within(pageList).getByText('1').className).toContain('page-link active');
        expect(previousLink.className).toContain('disabled');
        expect(nextLink.className).toContain('disabled');
        expect(screen.getByTestId('pagination')).toBeInTheDocument();
        expect(screen.getByTestId('pagination-page-info')).toBeInTheDocument();
        expect(screen.getByTestId('pagination-page-size-records')).toBeInTheDocument();
    });

    it('should go to the next page when click on next link', () => {
        expect.assertions(2);

        const handleClickNextPage = jest.fn();
        act(() => {
            render(<Pagination size={5} currentPage={1} totalCount={10} onPageChange={handleClickNextPage} />);
        });
        const nextLink = within(screen.getByTestId('pagination-next-link')).getByRole('button');

        fireEvent.click(nextLink);
        expect(handleClickNextPage).toHaveBeenCalledTimes(1);
        expect(handleClickNextPage).toHaveBeenCalledWith(2);
    });

    it('should go to the previous page when click on previous link', () => {
        expect.assertions(2);

        const handleClickPreviousPage = jest.fn();
        act(() => {
            render(<Pagination size={5} currentPage={2} totalCount={100} onPageChange={handleClickPreviousPage} />);
        });
        const previousLink = within(screen.getByTestId('pagination-previous-link')).getByRole('button');

        fireEvent.click(previousLink);
        expect(handleClickPreviousPage).toHaveBeenCalledTimes(1);
        expect(handleClickPreviousPage).toHaveBeenCalledWith(1);
    });

    it('should change page size and set current page to 1', () => {
        expect.assertions(4);

        const handleChangePageSize = jest.fn();
        act(() => {
            render(<Pagination size={5} currentPage={2} totalCount={100} onPageSizeChange={handleChangePageSize} />);
        });
        const pageSizeSelect = screen.getByTestId('pagination-page-size-records');

        fireEvent.change(pageSizeSelect, { target: { value: 10 } });

        const pageList = screen.getByTestId('pagination-page-list');
        expect(handleChangePageSize).toHaveBeenCalledTimes(1);
        expect(handleChangePageSize).toHaveBeenCalledWith(10);
        expect(screen.getByTestId('pagination-page-info').innerHTML).toBe('1-10 of 100');
        expect(within(pageList).getByText('1').className).toContain('page-link active');
    });

    afterAll(cleanup);
});
