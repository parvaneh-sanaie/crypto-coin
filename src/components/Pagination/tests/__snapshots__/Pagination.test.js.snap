// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Should be match to the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container flex-fill "
        data-testid="pagination"
      >
        <div
          class="row"
        >
          <div
            class="col-xl-3 col-md-2 d-flex flex-fill justify-content-start"
          >
            <label
              data-testid="pagination-page-info"
            >
              1
              -
              1
               of
               
              1
            </label>
          </div>
          <div
            class="col-xl-6 col-md-8 d-flex flex-fill justify-content-center"
          >
            <nav>
              <ul
                class="pagination"
                data-testid="pagination-page-list"
              >
                <li
                  class="page-item disabled"
                  data-testid="pagination-previous-link"
                >
                  <button
                    class="page-link"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                    >
                      «
                    </span>
                  </button>
                </li>
                <li
                  class="page-item"
                >
                  <span
                    class="page-link active"
                  >
                    1
                  </span>
                </li>
                <li
                  class="page-item disabled"
                  data-testid="pagination-next-link"
                >
                  <button
                    class="page-link"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                    >
                      »
                    </span>
                  </button>
                </li>
              </ul>
            </nav>
          </div>
          <div
            class="col-xl-3 col-md-2 d-flex flex-fill justify-content-end"
          >
            <label>
              Rows per page:
              <select
                data-testid="pagination-page-size-records"
              >
                <option
                  value="5"
                >
                  5
                </option>
                <option
                  value="10"
                >
                  10
                </option>
                <option
                  value="20"
                >
                  20
                </option>
                <option
                  value="50"
                >
                  50
                </option>
                <option
                  value="100"
                >
                  100
                </option>
              </select>
            </label>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container flex-fill "
      data-testid="pagination"
    >
      <div
        class="row"
      >
        <div
          class="col-xl-3 col-md-2 d-flex flex-fill justify-content-start"
        >
          <label
            data-testid="pagination-page-info"
          >
            1
            -
            1
             of
             
            1
          </label>
        </div>
        <div
          class="col-xl-6 col-md-8 d-flex flex-fill justify-content-center"
        >
          <nav>
            <ul
              class="pagination"
              data-testid="pagination-page-list"
            >
              <li
                class="page-item disabled"
                data-testid="pagination-previous-link"
              >
                <button
                  class="page-link"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    «
                  </span>
                </button>
              </li>
              <li
                class="page-item"
              >
                <span
                  class="page-link active"
                >
                  1
                </span>
              </li>
              <li
                class="page-item disabled"
                data-testid="pagination-next-link"
              >
                <button
                  class="page-link"
                  type="button"
                >
                  <span
                    aria-hidden="true"
                  >
                    »
                  </span>
                </button>
              </li>
            </ul>
          </nav>
        </div>
        <div
          class="col-xl-3 col-md-2 d-flex flex-fill justify-content-end"
        >
          <label>
            Rows per page:
            <select
              data-testid="pagination-page-size-records"
            >
              <option
                value="5"
              >
                5
              </option>
              <option
                value="10"
              >
                10
              </option>
              <option
                value="20"
              >
                20
              </option>
              <option
                value="50"
              >
                50
              </option>
              <option
                value="100"
              >
                100
              </option>
            </select>
          </label>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
